### PR TITLE
Add Google Protocol Buffers serialization support

### PR DESCRIPTION
**Pull request by gierlachg**
Added Google Protocol Buffers serialization support in form of a `StreamSerializer` adapter.
It can be registered on a job-level or instantiated and registered for the client/cluster.

Checklist
- [x] Tags Set
- [x] Milestone Set
- [x] New public APIs have `@Nonnull/@Nullable` annotations
- [x] New public APIs have `@since` tags in Javadoc

### CODE REVIEW

File: extensions/protobuf/src/main/java/com/hazelcast/jet/protobuf/ProtoStreamSerializer.java [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+
+    /**
+     * Creates Google Protocol Buffers v3 serializer.
+     *
+     * @param typeId unique type id of serializer
+     * @param clazz  class of {@link GeneratedMessageV3} handled by this
+     *               serializer
+     */
+    public ProtoStreamSerializer(int typeId, @Nonnull Class<T> clazz) {
+        checkTrue(GeneratedMessageV3.class.isAssignableFrom(clazz), clazz.getName() + " is not supported");
```
Comment (cangencer) [inline extensions/protobuf/src/main/java/com/hazelcast/jet/protobuf/ProtoStreamSerializer.java]: could we explain why? i.e. it needs to be a certain type etc.

File: extensions/protobuf/src/main/java/com/hazelcast/jet/protobuf/ProtoStreamSerializer.java [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+    }
+
+    @Override
+    public void write(ObjectDataOutput out, T object) throws IOException {
+        out.writeByteArray(object.toByteArray());
+    }
+
+    @Override
+    public T read(ObjectDataInput in) throws IOException {
+        return parser.parseFrom(in.readByteArray());
```
Comment (cangencer) [inline extensions/protobuf/src/main/java/com/hazelcast/jet/protobuf/ProtoStreamSerializer.java]: is `parser` thread safe?

File: hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/serialization/SerializerFactory.java [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+/**
+ * A serializer factory facade that discovers serializer factories by reading:
+ * <ul>
+ *     <li>"META-INF/services/com.hazelcast.jet.protobuf.SerializerFactoryHook"
+ *     for Google Protocol Buffers v3 serializer factory</li>
+ * </ul>
+ * <p>
+ * This system is meant to be internal code and is subject to change at any time.
+ */
+public final class SerializerFactory {
```
Comment (cangencer) [inline hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/serialization/SerializerFactory.java]: how is this different than `SerializerHook`? It seems like we created lot of new interfaces, without a concrete reason.

File: extensions/protobuf/src/main/java/com/hazelcast/jet/protobuf/ProtoStreamSerializer.java (hunk 2 above)
Comment (gierlachg) [reply extensions/protobuf/src/main/java/com/hazelcast/jet/protobuf/ProtoStreamSerializer.java]: Yes:
> The implementation should be stateless and thread-safe.

File: extensions/protobuf/src/main/java/com/hazelcast/jet/protobuf/ProtoSerializer.java [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+
+    @Override
+    public T read(ObjectDataInput in) throws IOException {
+        return parser.parseFrom(in.readByteArray());
+    }
+
+    /**
+     * An utility method that creates an anonymous {@link ProtoSerializer}.
+     */
+    public static <T extends GeneratedMessageV3> ProtoSerializer<T> from(@Nonnull Class<T> clazz, int typeId) {
```
Comment (cangencer) [inline extensions/protobuf/src/main/java/com/hazelcast/jet/protobuf/ProtoSerializer.java]: missing annotation on return type. Could also explain the parameters?

File: extensions/protobuf/src/main/java/com/hazelcast/jet/protobuf/ProtobufSerializer.java [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+import static com.hazelcast.jet.impl.util.ExceptionUtil.sneakyThrow;
+
+/**
+ * Adapts a generated <a
+ * href="https://developers.google.com/protocol-buffers/docs/proto3">
+ * Google Protocol Buffers v3</a> serializer class to the Hazelcast {@link
+ * StreamSerializer} interface.
+ * <p>
+ * To learn how to use this class, visit Jet's <a
+ * href="https://hazelcast-jet.io/docs/api/serialization#google-protocol-buffers">
```
Comment (cangencer) [inline extensions/protobuf/src/main/java/com/hazelcast/jet/protobuf/ProtobufSerializer.java:39]: should be jet-start.sh

File: extensions/protobuf/src/main/java/com/hazelcast/jet/protobuf/ProtobufSerializerHook.java [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+/**
+ * Simplifies using the {@linkplain SerializerHook Hazelcast Serializer
+ * Hook mechanism} to register a Google Protocol Buffers serializer with
+ * the Hazelcast Jet cluster. The serialized class can then be used in
+ * Hazelcast data structures and Jet jobs.
+ * <p>
+ * To use it, create a JAR containing the serializer hook class and
+ * {@code META-INF/services/com.hazelcast.SerializerHook}, a plain-text
+ * file containing its fully qualified name. Visit Jet's <a
+ * href="https://hazelcast-jet.io/docs/api/serialization#google-protocol-buffers">
```
Comment (cangencer) [inline extensions/protobuf/src/main/java/com/hazelcast/jet/protobuf/ProtobufSerializerHook.java:32]: same as before

**Pull request merged**